src/ctap2/mod.rs

```diff
@@ -1245,51 +1361,69 @@ pub(crate) fn credential_management(
 
 pub(crate) fn configure_authenticator(
     dev: &mut Device,
+    puat_result: Option<PinUvAuthResult>,
     cfg_subcommand: AuthConfigCommand,
     status: Sender<crate::StatusUpdate>,
     callback: StateCallback<crate::Result<crate::ManageResult>>,
     alive: &dyn Fn() -> bool,
-) {
+) -> bool {
     let mut authcfg = AuthenticatorConfig::new(cfg_subcommand);
     let mut skip_uv = false;
     let authinfo = match dev.get_authenticator_info() {
         Some(i) => i.clone(),
         None => {
             callback.call(Err(HIDError::DeviceNotInitialized.into()));
-            return;
+            return false;
         }
     };
 
     if authinfo.options.authnr_cfg != Some(true) {
         callback.call(Err(AuthenticatorError::HIDError(
             HIDError::UnsupportedCommand,
         )));
-        return;
+        return false;
     }
 
+    let mut skip_puap = false;
+    let mut pin_uv_auth_result = puat_result
+        .clone()
+        .unwrap_or(PinUvAuthResult::NoAuthRequired);
+    match puat_result {
+        Some(PinUvAuthResult::SuccessGetPinToken(t))
+        | Some(PinUvAuthResult::SuccessGetPinUvAuthTokenUsingUvWithPermissions(t))
+        | Some(PinUvAuthResult::SuccessGetPinUvAuthTokenUsingPinWithPermissions(t)) => {
+            skip_puap = true;
+            if authcfg.set_pin_uv_auth_param(Some(t)).is_err() {
+                return false;
+            }
+        }
+        _ => {}
+    }
     let mut pin = None;
     while alive() {
         // We can use the AuthenticatorConfiguration-command only in two cases:
         // 1. The device also supports the uv_acfg-permission (otherwise we can't establish a PUAP)
         // 2. The device is NOT protected by PIN/UV (yet). This allows organizations to configure
         //    the token, before handing them out.
         // If authinfo.options.uv_acfg is not supported, this will return UnauthorizedPermission
-        let pin_uv_auth_result = match determine_puap_if_needed(
-            &mut authcfg,
-            dev,
-            skip_uv,
-            PinUvAuthTokenPermission::AuthenticatorConfiguration,
-            UserVerificationRequirement::Preferred,
-            &status,
-            &callback,
-            alive,
-            &mut pin,
-        ) {
-            Ok(r) => r,
-            Err(()) => {
-                return;
-            }
-        };
+        if !skip_puap {
+            pin_uv_auth_result = match determine_puap_if_needed(
+                &mut authcfg,
+                dev,
+                skip_uv,
+                PinUvAuthTokenPermission::AuthenticatorConfiguration,
+                UserVerificationRequirement::Preferred,
+                &status,
+                &callback,
+                alive,
+                &mut pin,
+            ) {
+                Ok(r) => r,
+                Err(()) => {
+                    return false;
```

Comment:
And here?

Reply:
No, `determine_puap_if_needed()` already does the callback for us. 
This is indeed not pretty. We could probably stop handing in `callback` and instead return the actual `AuthenticatorError`, and `unwrap_result!()` here (and in all other places). 
But not sure if we want to inflate this PR even more. Maybe in a follow-up PR?

Reply:
Yeah, let's not make that change here.